apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,131 @@
+        internal long RowCount { get; private set; }
+
+        /// <summary>
+        /// Constructor for <see cref="BigtableReadRowsRequestManager"/>.
+        /// </summary>
+        /// <param name="originalRequest">
+        /// Original ReadRowsRequest containing all of the parameters of the API call.
+        /// </param>
+        internal BigtableReadRowsRequestManager(ReadRowsRequest originalRequest) => _originalRequest = originalRequest;
+
+        /// <summary>
+        /// Update last found Rowkey
+        /// </summary>
+        internal BigtableByteString LastFoundKey { get; set; }
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        internal void IncrementRowCount(int count) => RowCount += count;
+
+        /// <summary>
+        /// Builds and returns updated subrequest that excludes all rowKeys that have already been found.
+        /// </summary>
+        internal ReadRowsRequest BuildUpdatedRequest()
+        {
+            ReadRowsRequest newReadRowsRequest = new ReadRowsRequest
+            {
+                // Transfer TableName
+                TableNameAsTableName = _originalRequest.TableNameAsTableName,
+                // Transfer RowFilter.
+                Filter = _originalRequest.Filter,
+                // Remove received rowKeys form Rows.
+                Rows = FilterRows()
+            };
+
+            // If the row limit is set, update it.
+            if (_originalRequest.RowsLimit != 0)
+            {
+                long rowsRemaining = _originalRequest.RowsLimit - RowCount;
+                Debug.Assert(rowsRemaining > 0, "The remaining number of rows must be greater than 0.");
+                newReadRowsRequest.RowsLimit = Math.Max(1, rowsRemaining);
+            }
+
+            return newReadRowsRequest;
+        }
+
+        /// <summary>
+        /// Removes received rows from the RowSet.
+        /// </summary>
+        /// <returns>
+        /// New RowSet with not received rows.</returns>
+        private RowSet FilterRows()
+        {
+            RowSet originalRows = _originalRequest.Rows;
+
+            if (LastFoundKey.Value.IsEmpty)
+            {
+                return originalRows;
+            }
+
+            RowSet newRowSet = new RowSet
+            {
+                RowKeys = {originalRows.RowKeys.Where(key => !StartKeyIsAlreadyRead(key))}
+            };
+
+            foreach (RowRange rowRange in originalRows.RowRanges)
+            {
+                RowRange.EndKeyOneofCase endKeyOneofCase = rowRange.EndKeyCase;
+                if (endKeyOneofCase == RowRange.EndKeyOneofCase.EndKeyClosed &&
+                    EndKeyIsAlreadyRead(rowRange.EndKeyClosed)
+                    || endKeyOneofCase == RowRange.EndKeyOneofCase.EndKeyOpen &&
+                    EndKeyIsAlreadyRead(rowRange.EndKeyOpen))
+                {
+                    continue;
+                }
+
+                RowRange newRange = rowRange.Clone();
```

Comment:
You don't need to do this until you get into the if statement below (but its fine to do it here if you're really trying to be safe about the ranges not being shared). Also, do any of the tests fail if this `Clone` call is removed? If not, it would be good to have one.

Reply:
If the current `rowRange` is not skipped by the first if statement on line 100 then, modified or not by the the second if statement, it will still be added to `newRowSet.RowRanges` on line 119

Reply:
I realize that, but it may be safe to assume they can be shared if others play nicely and don't modify the ranges in an existing request. So we could defer cloning until we know we're going to modify it. But my point was that if you want to be safe about things and prevent sharing altogether, that is fine. Either way, a test which fails without `Clone` being called should be added, or an assert should be added to one of the existing tests that modifies ranges to make sure the original request has the original range values.